### PR TITLE
[Feat] online audio output support for diffusion-based audio models

### PR DESCRIPTION
**Pull request by RuixiangMa**
## Purpose

Add audio output support for diffusion-based models (Stable Audio) in online serving. Audio responses are returned via /v1/chat/completions with audio_url content in base64 WAV format

## Test Result

```
curl -s http://localhost:8004/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "stabilityai/stable-audio-open-1.0",
    "messages": [
      {"role": "user", "content": "The sound of a dog barking"}
    ],
    "extra_body": {
      "num_inference_steps": 50,
      "guidance_scale": 7.0,
      "audio_start_in_s": 0.0,
      "audio_end_in_s": 10.0,
      "seed": 42
    }
  }' | jq -r '.choices[0].message.content[0].audio_url.url' | cut -d',' -f2- | base64 -d > output.wav
```

[output.wav](https://github.com/user-attachments/files/26436481/output.wav)

E2E:

tests/e2e/online_serving/test_stable_audio_online.py::test_stable_audio_online[single_card_001] PASSED              [100%]

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 1]
```diff
@@ -2213,6 +2215,26 @@ async def _create_diffusion_chat_completion(
...
+            elif hasattr(self._diffusion_engine, "get_diffusion_od_config"):
+                od_config = self._diffusion_engine.get_diffusion_od_config()
+
+            if od_config:
+                model_class_name = getattr(od_config, "model_class_name", None)
+
+            supports_audio = False
+            if model_class_name:
+                model_cls = DiffusionModelRegistry._try_load_model_cls(model_class_name)
+                supports_audio = model_cls is not None and getattr(model_cls, "support_audio_output", False)
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_chat.py]: `_try_load_model_cls` is a private method on vllm's `_ModelRegistry` — it could be removed/renamed upstream at any time. Since `SupportAudioOutput` is a `@runtime_checkable` Protocol, consider doing the `isinstance` check once at model-load time and storing a flag on `od_config` instead of dynamically importing the class on every request.

**Review by lishunyang12 — COMMENTED** [review]
left a few comments on the serving_chat changes

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -2237,8 +2259,55 @@ async def _create_diffusion_chat_completion(
...
+                    audio_tensor = audio_tensor.float().detach().cpu().numpy()
+                else:
+                    if isinstance(audio_data, list):
+                        audio_tensor = np.concatenate(audio_data, axis=-1)
+                    else:
+                        audio_tensor = audio_data
+
+                sample_rate = 24000
+                if model_class_name == "StableAudioPipeline":
+                    sample_rate = 44100
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Hardcoding sample rates per pipeline class name doesn't scale. The pipeline already knows its own rate — store it on `DiffusionOutput` or `od_config` so this doesn't need a new branch for every future audio model.

```suggestion
                sample_rate = getattr(od_config, "sample_rate", 44100)
```

File: examples/online_serving/text_to_audio/README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      "num_inference_steps": 100,
+      "guidance_scale": 7.0,
+      "audio_start_in_s": 0.0,
+      "audio_end_in_s": 10.0,
+      "seed": 42
+    }
+  }' | jq -r '.choices[0].message.content[0].audio_url.url' | cut -d',' -f2- | base64 -d > output.wav
+```
+
+### Method 2: Using OpenAI Python SDK
```
Comment (lishunyang12) [inline examples/online_serving/text_to_audio/README.md:34]: Port mismatch — curl uses 8091 but this says 8004.

```suggestion
client = OpenAI(base_url="http://localhost:8091/v1", api_key="none")
```

File: examples/online_serving/text_to_audio/README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "num_inference_steps": 100,
+        "guidance_scale": 7.0,
+        "audio_start_in_s": 0.0,
+        "audio_end_in_s": 10.0,
+        "seed": 42,
+    },
+)
+
+audio_url = response.choices[0].message.content[0].audio_url.url
+_, b64_data = audio_url.split(",", 1)
```
Comment (lishunyang12) [inline examples/online_serving/text_to_audio/README.md:55]: Empty code block — either fill it in or remove the section.

File: tests/e2e/online_serving/test_stable_audio_online.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+import pytest
+import soundfile as sf
+
+from tests.conftest import OmniServer, OmniServerParams, OpenAIClientHandler, dummy_messages_from_mix_data
+from tests.utils import hardware_test
+
+MODEL_PATH = os.environ.get(
+    "STABLE_AUDIO_MODEL_PATH",
+    "linyueqian/stable_audio_random",
```
Comment (alex-jw-brooks) [inline tests/e2e/online_serving/test_stable_audio_online.py]: can you remove this env var? since the tests should generally use the same model set and we don't do this for other models atm

File: tests/e2e/online_serving/test_stable_audio_online.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import base64
```
Comment (alex-jw-brooks) [inline tests/e2e/online_serving/test_stable_audio_online.py:4]: IMO since this test is on a tiny model and to mostly validate the serving endpoint, it may be more clear to put it under `tests/entrypoints/openai_api/...` since it can be hard to search through model specific named tests to find things.

cc @fhfuih @yenuo26 in case this it's overlapping with the test refactors / there is a more clear future plan for this

File: vllm_omni/entrypoints/openai/serving_chat.py (hunk 2 above)
Comment (alex-jw-brooks) [reply vllm_omni/entrypoints/openai/serving_chat.py]: I think we should add `sample_rate` as a property to `SupportAudioOutput` and move this default there, or require pipelines implementing the protocol to set it - then we can read it once at load time as well, and avoid having a separate fallback default value for online serving that may not be used in the offline case

File: vllm_omni/entrypoints/openai/serving_chat.py
Comment (alex-jw-brooks) [inline vllm_omni/entrypoints/openai/serving_chat.py:2521]: Nit - some of these comments are outdated - it would also be nice to separate some of the image / audio output handling if possible since this function is pretty long

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 7]
```diff
@@ -2237,8 +2259,55 @@ async def _create_diffusion_chat_completion(
...
+                    }
+                ]
+
+                content = audio_url_content
+
+                logger.info(
+                    "Diffusion chat completed for request %s: audio output",
+                    request_id,
+                )
+            elif not image_contents:
```
Comment (alex-jw-brooks) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Is the intention that image / audio generation are mutually exclusive, and if so maybe we should make it more clear unless I missed the docs somewhere - if we end up with some model that generates both image and audio in the future, only the audio part would be returned

**Comment by hsliuustc0106** [comment]
will #2387 use this feature?

**Comment by RuixiangMa** [comment]
> will #2387 use this feature?

Yes, that was the intention(tested). In practice, this should be usable for diffusion-based audio models that return waveform outputs, Stable Audio is the concrete case I have tested in the pr.

File: vllm_omni/entrypoints/openai/serving_chat.py (hunk 1 above)
Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/serving_chat.py]: Now using od_config.supports_audio_output populated at engine init time — no registry lookup per request.

File: vllm_omni/entrypoints/openai/serving_chat.py (hunk 2 above)
Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/serving_chat.py]: > Hardcoding sample rates per pipeline class name doesn't scale. The pipeline already knows its own rate — store it on `DiffusionOutput` or `od_config` so this doesn't need a new branch for every future audio model.

Fixed. Now reads from od_config.audio_sample_rate (set to 44100 for Stable Audio at init).

Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/serving_chat.py]: > I think we should add `sample_rate` as a property to `SupportAudioOutput` and move this default there, or require pipelines implementing the protocol to set it - then we can read it once at load time as well, and avoid having a separate fallback default value for online serving that may not be used in the offline case

Added to SupportAudioOutput Protocol and pipeline implementation.

File: examples/online_serving/text_to_audio/README.md (hunk 3 above)
Comment (RuixiangMa) [reply examples/online_serving/text_to_audio/README.md:34]: fixed!

File: examples/online_serving/text_to_audio/README.md (hunk 4 above)
Comment (RuixiangMa) [reply examples/online_serving/text_to_audio/README.md:55]: fixed

File: tests/e2e/online_serving/test_stable_audio_online.py (hunk 5 above)
Comment (RuixiangMa) [reply tests/e2e/online_serving/test_stable_audio_online.py]: removed it

File: vllm_omni/entrypoints/openai/serving_chat.py (hunk 7 above)
Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/serving_chat.py]: yes, now returns both: content = image_contents + audio_url_content.

File: tests/e2e/online_serving/test_stable_audio_online.py (hunk 6 above)
Comment (yenuo26) [reply tests/e2e/online_serving/test_stable_audio_online.py:4]: I totally agree with your point. In addition, we are adding jobs to run the UT test cases that require CUDA resources(https://github.com/vllm-project/vllm-omni/pull/2449). I believe it would be more suitable to place them there.

**Review by lishunyang12 — COMMENTED** [review]
## Review: Online audio output support for diffusion-based audio models

Overall this is a solid feature addition that follows the existing patterns for image output. I have a few concerns, mainly around code duplication and robustness, that I think should be addressed before merge.

---

### 1. Major: `_enrich_client_od_config` duplicates `OmniDiffusionConfig.enrich_config()` (stage_diffusion_client.py)

The new `_enrich_client_od_config()` function in `stage_diffusion_client.py` is a near-exact copy of `OmniDiffusionConfig.enrich_config()` in `data.py`, plus the new `audiodit` branch that was also added to `stage_diffusion_proc.py`. Having three copies of the same config-enrichment logic is a maintenance hazard.

The `StageDiffusionProc._enrich_config()` already calls `self._od_config.enrich_config()` (one line). The client-side code should do the same:

```python
def _enrich_client_od_config(od_config):
    od_config.enrich_config()
```

Then the `audiodit` branch should only be added once, in `OmniDiffusionConfig.enrich_config()`. This also avoids the risk of these copies drifting apart over time.

### 2. Major: `_populate_audio_output_metadata` duplicates logic already in `DiffusionEngine.__init__`

The PR adds audio-metadata population in two places:
- `DiffusionEngine.__init__` (diffusion_engine.py lines ~92-95)
- `_populate_audio_output_metadata` (stage_diffusion_client.py)

Both do the same `getattr(model_cls, "support_audio_output", False)` etc. Consider having a single source of truth, e.g., a method on `OmniDiffusionConfig` itself, or at least a shared helper.

### 3. Medium: Fragile config discovery in `serving_chat.py`

```python
if hasattr(self._diffusion_engine, "od_config"):
    od_config = self._diffusion_engine.od_config
elif hasattr(self._diffusion_engine, "get_diffusion_od_config"):
    od_config = self._diffusion_engine.get_diffusion_od_config()
```

`self._diffusion_engine` is typed as `AsyncOmni | None`, and `AsyncOmni` already has `get_diffusion_od_config()`. The first `hasattr` branch matching `od_config` directly would only fire if the engine happened to be some other type. This dual-fallback is confusing and suggests the typing is wrong or there is an untested code path. I would recommend using only `get_diffusion_od_config()` (which already handles the attribute-chain walking internally) and removing the `hasattr(... "od_config")` branch to keep it clear.

### 4. Medium: `audio_data` can be fetched even when `supports_audio` is False

```python
if supports_audio:
    audio_data = result.multimodal_output.get("audio")
```

But if `result.multimodal_output` does not have a `.get()` method (e.g., it is `None`), this will raise `AttributeError`. Add a guard:

```python
if supports_audio and hasattr(result, "multimodal_output") and result.multimodal_output:
    audio_data = result.multimodal_output.get("audio")
```

### 5. Minor: Class attributes placed before the docstring (pipeline_stable_audio.py)

```python
class StableAudioPipeline(nn.Module, SupportAudioOutput, DiffusionPipelineProfilerMixin):
    sample_rate = 44100
    audio_channel_first = True
    """
    Pipeline for text-to-audio generation using Stable Audio Open.
    ...
    """
```

The docstring should come immediately after the `class` line, before any attribute assignments. As written, tools that extract docstrings (Sphinx, `help()`) will not pick it up. Move the two attributes below the docstring.

### 6. Minor: Logging only for audio, not image+audio mixed case

```python
logger.info(
    "Diffusion chat completed for request %s: audio output",
    request_id,
)
```

This log line is inside the `if audio_data is not None:` block but outside any check for images. If a future model produces both images and audio, the log would only mention audio. Consider moving the log after the combined content assembly or making it reflect what was actually produced.

### 7. Nit: Documentation duplication

`docs/user_guide/examples/online_serving/text_to_audio.md` and `examples/online_serving/text_to_audio/README.md` contain nearly identical content. If the docs page already includes the example files via `--8<--` directives, the README could be much shorter (just a pointer to the docs page) to avoid maintaining two copies.

### 8. Test coverage

The e2e test looks good -- it validates WAV decoding, sample rate, and stereo channels. Consider adding a negative test case (e.g., invalid `audio_end_in_s < audio_start_in_s`) to verify error handling in the online path.

---

**Summary**: The feature works and the test demonstrates it. The main concern is the three-way duplication of config enrichment logic, which should be consolidated before this lands. The other items are moderate to minor.
